carla/recourse_methods/catalog/cem/cem.py

```diff
@@ -0,0 +1,578 @@
+                    # failure, either multiply by 10 if no solution found yet
+                    #          or do binary search with the known upper bound
+                    Const_LB[batch_idx] = max(Const_LB[batch_idx], CONST[batch_idx])
+                    if Const_UB[batch_idx] < 1e9:
+                        CONST[batch_idx] = (
+                            Const_LB[batch_idx] + Const_UB[batch_idx]
+                        ) / 2
+                    else:
+                        CONST[batch_idx] *= 10
+
+        # return the best solution found
+        overall_best_attack = overall_best_attack[0]
+        return overall_best_attack.reshape((1,) + overall_best_attack.shape)
+
+    def counterfactual_search(self, instance):
+
+        orig_prob, orig_class, orig_prob_str = model_prediction(
+            self.catalog_model.raw_model, np.expand_dims(instance, axis=0)
+        )
+
+        target_label = orig_class
+        orig_sample, target = generate_data(instance, target_label)
+
+        # start the search
+        counterfactual = self.attack(orig_sample, target)
+
+        adv_prob, adv_class, adv_prob_str = model_prediction(
+            self.catalog_model.raw_model, counterfactual
+        )
+        delta_prob, delta_class, delta_prob_str = model_prediction(
+            self.catalog_model.raw_model, orig_sample - counterfactual
+        )
+
+        INFO = "[kappa:{}, Orig class:{}, Adv class:{}, Delta class: {}, Orig prob:{}, Adv prob:{}, Delta prob:{}".format(
+            self.kappa,
+            orig_class,
+            adv_class,
+            delta_class,
+            orig_prob_str,
+            adv_prob_str,
+            delta_prob_str,
+        )
+        print(INFO)
+
+        if np.argmax(
+            self.catalog_model.raw_model.predict(instance.reshape(1, -1))
+        ) != np.argmax(
+            self.catalog_model.raw_model.predict(counterfactual.reshape(1, -1))
+        ):
+            counterfactual = counterfactual
+        else:
+            counterfactual = counterfactual
+            counterfactual[:] = np.nan
+
+        return instance, counterfactual.reshape(-1)
+
+    def get_counterfactuals(self, factuals: pd.DataFrame):
+        """
+        Compute a certain number of counterfactuals per factual example.
+
+
+        Parameters
+        ----------
+        factuals : pd.DataFrame
+            DataFrame containing all samples for which we want to generate counterfactual examples.
+            All instances should belong to the same class.
+
+        Returns
+        -------
+
+        """
+        # drop targets
+        target_name = self.data.target
+        instances = factuals.drop(columns=[target_name])
+
+        # normalize and one-hot-encoding
+        instances = self.catalog_model.perform_pipeline(instances)
+        instances = instances[self.catalog_model.feature_input_order]
+
+        counterfactuals = []
+        times_list = []
+
+        for i in range(instances.values.shape[0]):
```

Comment:
As @indyfree mentioned in one of my PRs, we should use `iterrows()` for looping through dataframes.

https://stackoverflow.com/questions/16476924/how-to-iterate-over-rows-in-a-dataframe-in-pandas

Reply:
fixed.